taskchampion/taskchampion/src/server/cloud/gcp.rs

```diff
@@ -0,0 +1,392 @@
+                    }
+                    return self.next();
+                }
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use uuid::Uuid;
+
+    /// Make a service if `GCP_TEST_BUCKET` is set, as well as a function to put a unique prefix on
+    /// an object name, so that tests do not interfere with one another.
+    ///
+    /// Set up this bucket with a lifecyle policy to delete objects with age > 1 day. While passing
+    /// tests should correctly clean up after themselves, failing tests may leave objects in the
+    /// bucket.
+    ///
+    /// When the environment variable is not set, this returns false and the test does not run.
+    /// Note that the Rust test runner will still show "ok" for the test, as there is no way to
+    /// indicate anything else.
+    fn make_service() -> Option<(GcpService, impl Fn(&str) -> Vec<u8>)> {
+        let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
+            return None;
+        };
+        let prefix = Uuid::new_v4();
+        Some((GcpService::new(bucket).unwrap(), move |n: &_| {
+            format!("{}-{}", prefix.as_simple(), n).into_bytes()
+        }))
+    }
+
+    #[test]
+    fn put_and_get() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"foo").unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"foo".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn get_missing() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"data").unwrap();
+        svc.del(&pfx("testy")).unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del_missing() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn list() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        let mut names: Vec<_> = (0..20).map(|i| pfx(&format!("pp-{i:02}"))).collect();
+        names.sort();
+        // Create 20 objects that will be listed.
+        for n in &names {
+            svc.put(n, b"data").unwrap();
+        }
+        // And another object that should not be included in the list.
+        svc.put(&pfx("xxx"), b"data").unwrap();
+
+        let got_names: Vec<_> = svc.list(&pfx("pp-")).collect::<Result<_>>().unwrap();
+        let mut got_names: Vec<_> = got_names.into_iter().map(|oi| oi.name).collect();
```

Comment:
Shadowing seems a little weird here when the first variable encompasses more than "name".

```suggestion
        let got_objects: Vec<_> = svc.list(&pfx("pp-")).collect::<Result<_>>().unwrap();
        let mut got_names: Vec<_> = got_objects.into_iter().map(|oi| oi.name).collect();
```